### PR TITLE
[`LlamaTokenizerFast`] Refactor default llama

### LINKED ISSUE #28577: Inconsistent behavior between tokenizer and fast tokenizer
### System Info

- `transformers` version: 4.36.2
- Platform: Linux-4.18.0-193.6.3.el8_2.v1.4.x86_64-x86_64-with-glibc2.29
- Python version: 3.8.10
- Huggingface_hub version: 0.19.4
- Safetensors version: 0.4.0
- Accelerate version: 0.25.0
- Accelerate config:    not found
- PyTorch version (GPU?): 2.0.1+cu118 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: No

### Who can help?

@ArthurZucker 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

``` python
from transformers import AutoTokenizer
tokenizer = AutoTokenizer.from_pretrained("codellama/CodeLlama-7b-hf", trust_remote_code=True, use_fast=False)
fast_tokenizer = AutoTokenizer.from_pretrained("codellama/CodeLlama-7b-hf", trust_remote_code=True, use_fast=True)
prompt = "▁<PRE>//"
inputs = tokenizer(prompt, return_tensors="pt")
print(f"tokenizer ids: {inputs.input_ids}")
inputs = fast_tokenizer(prompt, return_tensors="pt")
print(f"fast tokenizer ids: {inputs.input_ids}")
```

This scripts will output:
```
tokenizer ids: tensor([[    1, 32007,   458]])
fast tokenizer ids: tensor([[    1, 32007,   849]])
```

In the `tokenizer.json` from the model folder, we can see:
```
"//": 458,
"▁//": 849,
```

Fast tokenizer probably ignores the `<PRE>` token, is it a correct behavior?

### Expected behavior

Fast tokenizer should be consistent with normal tokenizer.

### LINKED ISSUE #29617: `LlamaTokenizerFast` wrong `word_id` references based on batch encoding
### System Info

- `transformers` version: 4.39.0.dev0 (commit hash https://github.com/huggingface/transformers/commit/923733c22bf4d3cc6661c8cd3b730b275e9a938e)
- Platform: Linux-6.5.0-10022-tuxedo-x86_64-with-glibc2.35
- Python version: 3.10.12
- Huggingface_hub version: 0.20.3
- Safetensors version: 0.4.2
- Accelerate version: not installed
- Accelerate config: not found
- PyTorch version (GPU?): 2.2.1+cu121 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: no
- Using distributed or parallel set-up in script?: no

### Who can help?

@ArthurZucker 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

```python
from transformers import AutoTokenizer

# any llama (1) based tokenizer
llama_tokenizer = AutoTokenizer.from_pretrained('hf-internal-testing/llama-tokenizer')

sentence = 'Sampel to demonstrate the issue'

input_encoding = llama_tokenizer(sentence, add_special_tokens=True)
tokens = input_encoding.tokens()
word_ids = input_encoding.word_ids()

print(f'Tokens produced: {tokens}\nReferenced word ids: {word_ids}')
```
Prints: 
`Tokens produced: ['<s>', '▁S', 'amp', 'el', '▁to', '▁demonstrate', '▁the', '▁issue']`
`Referenced word ids: [None, 0, 0, 0, 0, 0, 0, 0]`

### Expected behavior

Given a fast tokenizer, batch encodings should be able to return the tokenized tokens as well as the corresponding `word_ids`. The first `None` is to be expected though (see this [example](https://github.com/huggingface/transformers/issues/12665#issuecomment-879161405)).

### LINKED ISSUE #29626: Inconsistency behaviours between LlamaTokenizer and LlamaTokenizerFast when lstrip=True
### System Info

ubuntu
transformers==4.39.0dev

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Refer to tokenization test:

https://github.com/huggingface/transformers/blob/9acce7de1cb8229304a467938ebb47727d60cdb2/tests/test_tokenization_common.py#L864-L881
```python
from transformers import AutoTokenizer
from transformers.tokenization_utils import AddedToken

tokenizer = AutoTokenizer.from_pretrained("/path/to/llama2", use_fast=False)

special_token = tokenizer.all_special_tokens[0]

text = special_token + " aaaaa bbbbbb low cccccccccdddddddd l " + special_token
text2 = special_token + " AAAAA BBBBBB low CCCCCCCCCDDDDDDDD l " + special_token

toks_before_adding = tokenizer.tokenize(text)  # toks before adding new_toks

new_toks = ["aaaaa bbbbbb", "cccccccccdddddddd", "AAAAA BBBBBB", "CCCCCCCCCDDDDDDDD"]

added = tokenizer.add_tokens([AddedToken(tok, lstrip=True, rstrip=True) for tok in new_toks])

toks_after_adding = tokenizer.tokenize(text)
toks_after_adding2 = tokenizer.tokenize(text2)

print(toks_after_adding, toks_after_adding2)
```
If `use_fast=False`, the output is:
```bash
['<s>', 'aaaaa bbbbbb', '▁low', 'cccccccccdddddddd', '▁l', '▁', '<s>'] ['<s>', 'AAAAA BBBBBB', '▁low', 'CCCCCCCCCDDDDDDDD', '▁l', '▁', '<s>']
```
Otherwise
```bash
['<s>', '▁', '▁aaaaa▁bbbbbb', '▁low', '▁cccccccccdddddddd', '▁l', '▁', '<s>'] ['<s>', '▁', '▁AAAAA▁BBBBBB', '▁low', '▁CCCCCCCCCDDDDDDDD', '▁l', '▁', '<s>']
```

### Expected behavior

Seems that `lstrip=True` did not take effect. When `use_fast=True`, the output should be the same as `use_fast=False`

### LINKED ISSUE #29868: Inconsistent tokenization between fast and slow tokenizers for sentencepiece user-defined tokens.
### System Info

- `transformers` version: 4.38.2
- Platform: Linux-5.4.0-125-generic-x86_64-with-glibc2.29
- Python version: 3.8.10
- Huggingface_hub version: 0.21.4
- Safetensors version: 0.4.2
- Accelerate version: not installed
- Accelerate config: not found
- PyTorch version (GPU?): 2.1.0a0+fe05266 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>

### Who can help?

@ArthurZucker 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

Fast tokenizers cannot view the added token as a whole but split it as many small parts.
I found the similar issue in the huggingface discussion (but with no solution) : https://discuss.huggingface.co/t/sentencepiece-user-defined-symbols-and-fast-tokenizers/52208
Here is a script to reproduce this issue:

```python
from transformers import AutoTokenizer
import transformers.utils.sentencepiece_model_pb2 as sp_model
import json

proto = sp_model.ModelProto()

llama_orgin_tokenizer_path = 'Llama-2-7b-hf' # this a local path
llama_tokenizer_with_added_token_path = 'Llama-2-7b-hf-with-added-token'

with open(f'{llama_orgin_tokenizer_path}/tokenizer.model', 'rb') as fin:
    proto_str = fin.read()
    proto.ParseFromString(proto_str)

# append a user-defined token
adding_token = 'aaabbbcc'
proto.pieces.append(sp_model.ModelProto.SentencePiece(piece=adding_token, score=0.0, type=4)) #type 4 means user defined
with open(f'{llama_tokenizer_with_added_token_path}/tokenizer.model', 'wb') as fout:
    fout.write(proto.SerializeToString())


with open(f'{llama_tokenizer_with_added_token_path}/tokenizer_config.json','w', encoding='utf-8') as fout:
    tokenizer_config = {
        "model_max_length":99999999999999,
        "tokenizer_class": "LlamaTokenizer"
    }
    json.dump(tokenizer_config, fout)



tokenizer_fast = AutoTokenizer.from_pretrained(llama_tokenizer_with_added_token_path, use_fast=True)
tokenizer_slow = AutoTokenizer.from_pretrained(llama_tokenizer_with_added_token_path, use_fast=False)

text = 'I love aaabbbcc.'
print('fast:', tokenizer_fast.tokenize(text))
print('slow:', tokenizer_slow.tokenize(text))

# The outputs:
# fast: ['▁I', '▁love', '▁aa', 'ab', 'bb', 'cc', '.']
# slow: ['▁I', '▁love', '▁', 'aaabbbcc', '.']
```

### Expected behavior

It should be consistent tokenization results between fast and slow tokenizers for **those added tokens.**

### LINKED ISSUE #29872: Llama tokenizers behaviours different with special tokens.
### System Info

- `transformers` version: 4.39.1
- Platform: Linux-3.10.0-1062.9.1.el7.x86_64-x86_64-with-glibc2.35
- Python version: 3.10.12
- Huggingface_hub version: 0.19.4
- Safetensors version: 0.4.1
- Accelerate version: 0.25.0
- Accelerate config:    not found
- PyTorch version (GPU?): 2.1.2+cu121 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed

### Who can help?

@ArthurZucker 

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

the code sample is :
```
from transformers import AutoTokenizer

tok = AutoTokenizer.from_pretrained("./llama2", use_fast=False)

# add three extra different type tokens:
tok.add_tokens(['|<im_start1>|'], special_tokens=False)
tok.add_tokens(['|<im_start2>|'], special_tokens=True)
tok.add_special_tokens({"pad_token": '|<im_start3>|'})

print(tok)
# output:
"""
LlamaTokenizer(name_or_path='./llama2/', vocab_size=32000, model_max_length=1000000000000000019884624838656, is_fast=False, padding_side='right', truncation_side='right', special_tokens={'bos_token': '<s>', 'eos_token': '</s>', 'unk_token': '<unk>', 'pad_token': '|<im_start3>|', 'additional_special_tokens': ['|<im_start2>|']}, clean_up_tokenization_spaces=False),  added_tokens_decoder={
        0: AddedToken("<unk>", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
        1: AddedToken("<s>", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
        2: AddedToken("</s>", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
        32000: AddedToken("|<im_start1>|", rstrip=False, lstrip=False, single_word=False, normalized=True, special=False),
        32001: AddedToken("|<im_start2>|", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
        32002: AddedToken("|<im_start3>|", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
}
"""

print(tok.tokenize("|<im_start1>|system"))
# ['▁', '|<im_start1>|', 'system']

print(tok.tokenize("|<im_start2>|system"))
# ['|<im_start2>|', 'system']

print(tok.tokenize("|<im_start3>|system"))
# ['|<im_start3>|', 'system']

print(tok.decode(tok("|<im_start1>|system").input_ids))
# '<s>  |<im_start1>| system'

print(tok.decode(tok("|<im_start2>|system").input_ids))
# '<s> |<im_start2>| system'

print(tok.decode(tok("|<im_start3>|system").input_ids))
# '<s>|<im_start3>|system'
```

We can see that only the last example of using '|<im_start3>|' could recover the original input, and there will be a space in the other two examples.
So it is confusing, and I don't know what is right? (maybe the |<im_start3>| is the expected behavior?)
- what is the difference of the following two methods:
 ```
    tok.add_tokens(['|<im_start2>|'], special_tokens=True)
    tok.add_special_tokens({"pad_token": '|<im_start3>|'})
 ```
- Why there are extra space in the above three tokens? 
- How to change the behaviors of '|<im_start1>|' and '|<im_start2>|' to the same as '|<im_start3>|'? 

Would you mind taking care of this issue? @ArthurZucker 
Thanks a lot. 

### Expected behavior

I expect the behaviors of '|<im_start3>|'.

### LINKED ISSUE #30416: Gemma's Tokenizer fails to split on spaces
### Preface

This is related to #29617, similar issue as described in there so this is kind of a placeholder. Should be fixed by #28881.

### Who can help?

@ArthurZucker 

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```python
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained('google/gemma-7b', use_auth=True)
sentence = 'Sampel to demonstrate the issue'

input_encoding = tokenizer(sentence, add_special_tokens=True)
tokens = input_encoding.tokens()
word_ids = input_encoding.word_ids()

print(f'Tokens produced: {tokens}\nReferenced word ids: {word_ids}')
```

### Expected behavior

Correct split(s) assigned.

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
```python
from transformers import LlamaTokenizerFast, AddedToken
tokenizer = LlamaTokenizerFast.from_pretrained("huggyllama/llama-7b", legacy=False, from_slow=True)
tokenizer.add_tokens([AddedToken("<REPR_END>", rstrip=True, lstrip=True)], special_tokens=False)
tokenizer.tokenize("<REPR_END>inform<s>. Hey.       .")
['<REPR_END>', 'in', 'form', '<s>', '.', '▁Hey', '.', '▁▁▁▁▁▁', '▁.']
tokenizer.tokenize("inform<s>. Hey.       .")
['in', 'form', '<s>', '.', '▁Hey', '.', '▁▁▁▁▁▁', '▁.']
```

This requires https://github.com/huggingface/tokenizers/pull/1476 to be merged. 
Finally FIXES ALL THE remaining tokenizer issues!

Fixes #29617 as well (will update Gemma one)
fixes #28577
fixes #29617
fixes #29626
fixes #29694
fixes #29868
fixes #29872
fixes #30416
enabled by https://github.com/huggingface/tokenizers/pull/1476

### CODE REVIEW

File: src/transformers/convert_slow_tokenizer.py [hunk 1]
```diff
@@ -1247,14 +1247,6 @@ def tokenizer(self, proto):
         return tokenizer
 
     def normalizer(self, proto):
```
Comment (ArthurZucker) [inline src/transformers/convert_slow_tokenizer.py:1249]:
```suggestion
    def normalizer(self, proto):
        return normalizers.Replace(pattern=" ", content="▁"),
```

**Comment by ArthurZucker** [comment]
Coming next release!

**Review by LysandreJik — APPROVED** [review]
Sounds great to me

File: src/transformers/models/llama/tokenization_llama_fast.py [hunk 2]
```diff
@@ -91,6 +91,29 @@ class LlamaTokenizerFast(PreTrainedTokenizerFast):
...
+        legacy (`bool`, *optional*):
+            Whether or not the `legacy` behavior of the tokenizer should be used. Legacy is before the merge of #24622
+            and #25224 which includes fixes to properly handle tokens that appear after special tokens. A simple
+            example:
+
+            - `legacy=True`:
+            ```python
+            >>> from transformers import T5Tokenizer
+
+            >>> tokenizer = T5Tokenizer.from_pretrained("t5-base", legacy=True)
```
Comment (pcuenca) [inline src/transformers/models/llama/tokenization_llama_fast.py:104]: nit: shouldn't this be a llama tokenizer example?

Comment (ArthurZucker) [reply src/transformers/models/llama/tokenization_llama_fast.py:104]: Yep 👍🏻

File: src/transformers/convert_slow_tokenizer.py [hunk 3]
```diff
@@ -105,7 +105,7 @@ def extract(self, vocab_scores=None) -> Tuple[Dict[str, int], List[Tuple]]:
 
         # there is a missing token in the vocab. We have to do this to support merges
         # "<0x09>" is the bytefallback for `\t`
-        vocab["\t"] = vocab.pop("<0x09>")
+        vocab["\t"] = vocab.get("<0x09>")
```
Comment (ArthurZucker) [inline src/transformers/convert_slow_tokenizer.py:108]: this does not change anything

File: src/transformers/convert_slow_tokenizer.py [hunk 4]
```diff
@@ -1276,7 +1276,7 @@ def vocab(self, proto):
         return vocab
 
     def pre_tokenizer(self, replacement, add_prefix_space):
-        return None
+        return pre_tokenizers.Metaspace(replacement=replacement, prepend_scheme="never", split=False)
```
Comment (ArthurZucker) [inline src/transformers/convert_slow_tokenizer.py:1279]: actually we can probably just use a `pre_tokenizers.Split()` here, will try that

File: src/transformers/convert_slow_tokenizer.py [hunk 5]
```diff
@@ -1329,7 +1329,7 @@ def tokenizer(self, proto):
                 "You're trying to run a `Unigram` model but you're file was trained with a different algorithm"
             )
         user_defined_symbols = [
-            AddedToken(token, normalized=False, special=False) for token in proto.trainer_spec.user_defined_symbols
+            AddedToken(token, normalized=True, special=False) for token in proto.trainer_spec.user_defined_symbols
```
Comment (ArthurZucker) [inline src/transformers/convert_slow_tokenizer.py:1332]: This is a must? Unless with only the split this works.

File: src/transformers/convert_slow_tokenizer.py (hunk 4 above)
Comment (ArthurZucker) [reply src/transformers/convert_slow_tokenizer.py:1279]: TLDR we want to get the word ids
